File: kotlin/app/src/main/kotlin/se/krois/aoc2021/Day10.kt
package se.krois.aoc2021.day10

fun solve() {
    var syntaxErrorScore = 0
    var completionScores = mutableListOf<Long>()
    while (true) {
        val input = readLine()?.trim() ?: break
        when (val result = check(input)) {
            is Corrupted -> syntaxErrorScore += result.score
            is Completion -> completionScores.add(result.score)
        }
    }
    completionScores.sort()
    println("syntax error score: $syntaxErrorScore")
    println("completion score: ${completionScores[completionScores.size / 2]}")
}

sealed class CheckResult

class Corrupted(val score: Int) : CheckResult()

class Completion(val score: Long) : CheckResult()

fun check(line: String): CheckResult {
    val stack = mutableListOf<Char>()
    for (c in line) {
        when (c) {
            '(', '[', '{', '<' -> stack.add(c)
            ')', ']', '}', '>' -> {
                val top = stack.removeLastOrNull()
                when (Pair(top, c)) {
                    Pair('(', ')'), Pair('[', ']'), Pair('{', '}'), Pair('<', '>') -> continue
                }
                when (c) {
                    ')' -> return Corrupted(3)
                    ']' -> return Corrupted(57)
                    '}' -> return Corrupted(1197)
                    '>' -> return Corrupted(25137)
                }
            }
        }
    }

    var score = 0L
    for (c in stack.reversed()) {
        score *= 5
        when (c) {
            '(' -> score += 1
            '[' -> score += 2
            '{' -> score += 3
            '<' -> score += 4
        }
    }
    return Completion(score)
}

File: rust/src/day10.rs
use std::io;

pub fn solve(reader: &mut dyn io::BufRead) -> io::Result<()> {
    let mut line = String::new();
    let mut syntax_error_score = 0;
    let mut completion_scores = Vec::new();
    loop {
        line.clear();
        if reader.read_line(&mut line)? <= 0 {
            break;
        }
        match check(line.trim()) {
            CheckResult::Corrupted(s) => syntax_error_score += s,
            CheckResult::Completion(s) => completion_scores.push(s),
        }
    }
    completion_scores.sort();
    println!("syntax error score: {}", syntax_error_score);
    println!(
        "completion score: {:?}",
        completion_scores[completion_scores.len() / 2]
    );
    Ok(())
}

enum CheckResult {
    Corrupted(i32),
    Completion(i64),
}

fn check(line: &str) -> CheckResult {
    let mut stack = Vec::new();
    for c in line.chars() {
        match c {
            '(' | '[' | '{' | '<' => stack.push(c),
            ')' | ']' | '}' | '>' => {
                let top = stack.pop();
                match (top, c) {
                    (Some('('), ')') | (Some('['), ']') | (Some('{'), '}') | (Some('<'), '>') => {}
                    (_, ')') => {
                        return CheckResult::Corrupted(3);
                    }
                    (_, ']') => {
                        return CheckResult::Corrupted(57);
                    }
                    (_, '}') => {
                        return CheckResult::Corrupted(1197);
                    }
                    (_, '>') => {
                        return CheckResult::Corrupted(25137);
                    }
                    _ => {}
                }
            }
            _ => {}
        }
    }
    let mut score: i64 = 0;
    for c in stack.iter().rev() {
        score *= 5;
        match c {
            '(' => score += 1,
            '[' => score += 2,
            '{' => score += 3,
            '<' => score += 4,
            _ => {}
        }
    }
    CheckResult::Completion(score)
}
